File: AoC day 2/rust/solution_second_part.rs
#!/usr/bin/env rust-script

fn main() {
	let input = std::fs::read_to_string("input").unwrap();

	let res = input
		.split_terminator("\n") // per evitare l'ultima riga vuota
		.map(|line| {
			let mut split = line.split(' ');
			
			( split.next().unwrap(), split.next().unwrap() )
		})
		.map(|(first, second)| {
			( 
				first.chars().next().unwrap() as u8 - 'A' as u8, 
				second.chars().next().unwrap() as u8 - 'X' as u8
			)
		})
		.map(|(first, second)| {
			(
				first, 
				(first + second + 2) % 3
			)
		})
		.map(|(first, second)| {
			let play_diff = (3 + second - first) % 3; // +3 per evitare numeri negativi
			
			let won_pts = match play_diff {
				0 => 3,
				1 => 6,
				2 => 0,
				_ => panic!("Invalid play diff"),
			};

			let play_pts = second + 1;

			(won_pts + play_pts) as u32
		})
		.sum::<u32>();
		
	println!("{res}");
}

File: AoC day 2/c/solution_second_part.c
//usr/bin/gcc solution_second_part.c && exec ./a.out

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>

int main(int argc, char** argv) {
	FILE *fp = NULL;
	
	fp = fopen("input", "r");
	if (fp == NULL)
		exit(1);

	char line[8], first, second, diff;
	uint32_t score = 0;
	while (!feof(fp)) {
		if (!fgets(line, sizeof line, fp))
			break;

		first = line[0] - 'A';
		second = line[2] - 'X';

		second = (first + second + 2) % 3;

		diff = (3 + second - first) % 3;

		if (diff == 0) {
			score += 3;
		} else if (diff == 1) {
			score += 6;
		}

		score += second + 1;
	}

	printf("%d\n", score);
}
